daffodil-lib/src/main/scala/org/apache/daffodil/api/Validator.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.api
+
+import org.xml.sax.ErrorHandler
+
+trait Validator {
+  def name(): String = getClass.getTypeName.toLowerCase
+
+  def checkArgs(args: Validator.Arguments): Either[String, Unit] = Right(())
+
+  def validateXML(document: java.io.InputStream, errHandler: ErrorHandler, args: Validator.Arguments): Unit
```

Comment:
I think rather than requiring validators to cache and hide the fact that validation has a compilation step in it, we should make the API allow for this compilation step explicitly. 

I think the validator you find via SPI should have an explicit compile(params) step that returns a CompiledValidator, and a CompiledValidator should have a validate(xmlInput, errorHandler) call on it. as well as be Serializable so it can be saved/restored as part of the runtime objects of a compiled DFDL schema binary. 

The validate(xmlInput, errorHandler) should be thread safe, and if it depends on underlying non-threadsafe libraries it should deal with using ThreadLocals so that the caller need not be aware of this.

Reply:
OK.  That can probably work.  I see how it relates to your comment above.  Will take a stab at this compiled idea tomorrow.

Reply:
Here is a sketch in another branch, if its close Ill move it over and we can advance it, just didnt want to blow this PR up if it was way off.

https://github.com/jw3/incubator-daffodil/commit/d9f7f4346e57e12aa2d14e24c5c5ca0581124a26

I like the compiled approach, it addresses a few pain points and it ends up being more flexible to allow non-spi usage for API clients as well.

Reply:
This approach is working out pretty well and goes along with discussion on the thread local usage, so pushed it to this branch for review